src/evaluation/evaluator.h

```diff
@@ -6,322 +6,348 @@
 #include "evaluation/pv_table.h"
 #include "evaluation/repetition.h"
 #include "evaluation/static_evaluation.h"
+
 #include "fmt/ranges.h"
 #include "movegen/move_types.h"
 #include "syzygy/syzygy.h"
 #include <atomic>
 #include <engine/zobrist_hashing.h>
 
 #include <chrono>
+#include <condition_variable>
+#include <future>
 #include <mutex>
 #include <thread>
 
 namespace evaluation {
 
-class Evaluator {
+enum SearchType {
+    Default,
+    NullSearch,
+};
+
+enum SearchState {
+    Idle,
+    Searching,
+    Kill
+};
+
+struct MoveResult {
+    uint64_t hash;
+    BitBoard board;
+};
+
+struct SearcherResult {
+    int32_t score;
+    uint8_t searcherId;
+    movegen::Move pvMove;
+    uint8_t searchedDepth;
+};
+
+// Wrapper around a static array for intuitive iteration
+template<typename T, uint8_t maxSize>
+class StackVector {
 public:
-    constexpr movegen::Move getBestMove(const BitBoard& board, std::optional<uint8_t> depthInput = std::nullopt)
-    {
-        m_startTime = std::chrono::system_clock::now();
+    uint8_t size() { return m_size; }
 
-        /*
-         * if a depth has been provided then make sure that we search to that depth
-         * timeout will be 10 minutes
-         */
-        if (depthInput.has_value()) {
-            m_endTime = m_startTime + std::chrono::minutes(10);
-        } else {
-            setupTimeControls(m_startTime, board);
-        }
+    T& at(uint8_t idx) { return m_entries.at(idx); }
 
-        m_hash = engine::generateHashKey(board);
+    void setSize(uint8_t size) { m_size = size; }
 
-        return scanForBestMove(depthInput.value_or(s_maxSearchDepth), board);
-    }
+    void incSize() { m_size++; }
 
-    constexpr std::optional<movegen::Move> getPonderMove() const
+    auto* begin() { return m_entries.begin(); }
+
+    auto* end() { return m_entries.begin() + m_size; }
+
+private:
+    std::array<T, maxSize> m_entries;
+    uint8_t m_size { 0 };
+};
+
+template<typename T, typename U, uint8_t maxSize>
+class StackMap {
+public:
+    StackMap() = default;
+
+    StackMap(std::initializer_list<std::pair<T, U>> init)
     {
-        if (m_moveOrdering.pvTable().size() > 1) {
-            return m_moveOrdering.pvTable()[1];
-        }
+        assert(init.size() < maxSize);
 
-        return std::nullopt;
+        for (auto& entry : init) {
+            m_entries.at(m_size++) = std::move(entry);
+        }
     }
 
-    /* TODO: ensure time handler is stopped as well - not just assumed stopped. This is racy */
-    constexpr void stop()
+    constexpr U& at(const T& ext_key)
     {
-        m_isStopped = true;
+        for (auto& [key, value] : m_entries) {
+            if (key == ext_key) {
+                return value;
+            }
+        }
+        throw std::out_of_range("No entry matching this key");
```

Comment:
We don't catch this anywhere - catch / throw is extremely slow so we should avoid using it